### PR TITLE
Add airborne state and ball-table collision

### PR DESCRIPTION
**Pull request by ekiefl**
## Adds:

- Airborne state
- Airborne equations of motion
- Math for the time until ball-table collision (quadratic)
- Ball-table collision resolution infrastructure
- An example ball-table collision model: frictionless inelastic
- Integration of ball-table collisions into the event-based simulator

## Changes:

- The math/geometry for all other collision types needs to be updated in some way or another, so these collisions have been turned off and should be incrementally turned off in future PRs.

## Trial it out:

I made a demo called `sandbox/airborne_demos.py` which we can add different scenarios to.

## Next up:

I think natural next steps (in future PRs) would be to start re-incorporating the other collisions.

### CODE REVIEW

**Comment by ekiefl** [comment]
@derek-mcblane This is ready for your review!

**Review by derek-mcblane — APPROVED** [review]
Awesome work, @ekiefl!  The demos were really exciting to see.

No major concerns.

It was probably easiest to just use the quadratic solver you already had, but for predicting collision time, we can always use the (b **+** sqrt(D))/(2a) solution if we make some precondition assumptions.  Basically "what goes up must come down."*

*Assumptions:
- ball starts above the table
- acceleration due to gravity is downward towards table

File: pooltool/evolution/event_based/solve.py [hunk 1]
```diff
@@ -191,6 +191,52 @@ def ball_ball_collision_time(
...
+    min_time = np.inf
+    for root in roots:
+        if np.isnan(root):
+            # This is an indirect test for whether the root is complex or not. This is
+            # because ptmath.roots.quadratic.solve returns nan if the root is complex.
+            continue
+
+        if root.real <= const.EPS:
+            # FIXME-3D Is this necessary?
+            continue
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply pooltool/evolution/event_based/solve.py:232]: I copied it from the nearby `ball_linear_cushion_collision_time` function. I don't think it's necessary... Removing them both.

[the remark this answers is not in the document]
Comment (ekiefl) [reply pooltool/evolution/event_based/solve.py:232]: Yikes, it seems like it *is* necessary. I think it has to do with some numerical stability issues. I've been through hell and back with numerical stability and root solving over the years, so I'm going to let sleeping dogs rest...

[the remark this answers is not in the document]
Comment (ekiefl) [reply pooltool/evolution/event_based/solve.py]: I applied your simplification of the ball-table collision time in this commit 751df1326a35ff048226ec5d72324dc1a2f66306 which let me remove the weird `const.EPS` term. 🥳

File: pooltool/physics/evolve/__init__.py [hunk 2]
```diff
@@ -171,9 +161,39 @@ def evolve_perpendicular_spin_component(
...
+
+    v = v_0.copy()
+    v[2] = v_0[2] - g * t
+
+    new_rvw = np.empty((3, 3), dtype=np.float64)
+    new_rvw[0, :] = r
+    new_rvw[1, :] = v
+    new_rvw[2, :] = w_0
+
+    return new_rvw
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply pooltool/physics/evolve/__init__.py]: 28fde4b1f00c1eb1322721ffebe5da1e0c232ead

File: pooltool/evolution/event_based/solve.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekiefl) [reply pooltool/evolution/event_based/solve.py]: I ended up refactoring the `get_airborne_time` fn to use your simplification, and then in `ball_table_collision_time` I just delegate to `get_airborne_time` (9baf75d5b750bc85081369bb6e3d7705601fc2cc)

**Comment by ekiefl** [comment]
Thanks @derek-mcblane, this was all really good feedback. Code looks a lot better now.

File: pooltool/evolution/event_based/solve.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (derek-mcblane) [reply pooltool/evolution/event_based/solve.py]: > Yikes, it seems like it _is_ necessary. I think it has to do with some numerical stability issues. I've been through hell and back with numerical stability and root solving over the years, so I'm going to let sleeping dogs rest...

Glad it's not just me!

[the remark this answers is not in the document]
Comment (derek-mcblane) [reply pooltool/evolution/event_based/solve.py]: I've run into a lot of issues with extraneous collisions predicted just after a collision.  I'm dealing with one right now trying to get arbitrarily rotated cylinder collisions working.  It clicked that this issue is probably what that `const.EPS` check is for here.

**Pull request merged**